### PR TITLE
Review

### PR DESCRIPTION
**Pull request by tiagoantunespt**
I'm opening this PR to allow us to add some comments/questions on the code.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,9 +3,21 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@material-ui/core": "^4.4.1",
+    "@material-ui/icons": "^4.4.1",
+    "axios": "^0.19.0",
+    "bootstrap": "^4.3.1",
+    "node-sass": "^4.12.0",
     "react": "^16.9.0",
+    "react-bootstrap": "^1.0.0-beta.12",
```
Comment (tiagoantunespt) [inline package.json:12]: Why did you include `react-bootstrap` and `bootstrap` libraries?

File: src/actions/employee.action.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import {employeesData} from '../data/employee-mock.data';
```
Comment (tiagoantunespt) [inline src/actions/employee.action.js]: What do you know about redux actions? Can you elaborate why did you add this functions in this `actions` folder?

File: src/components/Login/Login.jsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+import {connect} from 'react-redux';
+
+import {makeStyles} from '@material-ui/core/styles';
+import TextField from '@material-ui/core/TextField';
+import Button from '@material-ui/core/Button';
+
+import {ActionTypes} from '../../constants/Actions';
+import './Login.css';
+
+const useStyles = makeStyles(theme => ({
```
Comment (tiagoantunespt) [inline src/components/Login/Login.jsx:12]: You are using both jss styles (here) and `.css` files. What do you think about this two different ways to manage css?

File: src/reducers/employeeReducer.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        ...state,
+        status: null,
+        error: null
+      };
+    }
+    case ActionTypes.DELETE_EMPLOYEE_SUCCESS : {
+      return {
+        ...state,
+        employees: action.result,
+        status: 'Employee Deleted Successfully',
```
Comment (tiagoantunespt) [inline src/reducers/employeeReducer.js:47]: You added the messages to show in the UI here in the reducers. Can you elaborate another approach to do this?

File: src/reducers/employeeReducer.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  error: null,
+  selectedEmployee: {},
+  status: null
+};
+
+const employeeReducer = (state = initialState, action) => {
+  switch (action.type) {
+    case ActionTypes.FETCH_EMPLOYEES_SUCCESS : {
+      return {
+        ...state,
```
Comment (tiagoantunespt) [inline src/reducers/employeeReducer.js:14]: Can you explain why did you use the spread operator here? 

What if you did something like this?
```
state.status = null
return status
```

File: src/reducers/userReducer.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    case ActionTypes.DELETE_TOKEN: {
+      newState = {...state, loginInfo: {...state.loginInfo, token: null, user: null}, loginError: null};
+      localStorage.removeItem('login_token');
+      localStorage.removeItem('userLoggedIn');
+      newState.token = null;
+      return newState;
+    }
+    case ActionTypes.LOGOUT: {
+      newState = {...state, loginInfo: {...state.loginInfo, token: null, status: LOGGED_OUT, user: null}, loginError: null};
```
Comment (tiagoantunespt) [inline src/reducers/userReducer.js:34]: this line has 125 columns. Do you use any library (js linter, prettier, standardjs) to help you having the code automatically well formatted?

File: src/sagas/employeeSaga/delete.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import {takeLatest, put, call} from 'redux-saga/effects';
+import {ActionTypes} from '../../constants/Actions';
+import {deleteEmployeeById} from '../../actions/employee.action';
+
+function* deleteEmployee(action) {
+  yield put({type: ActionTypes.DELETE_EMPLOYEE_PENDING, id: action.id});
```
Comment (tiagoantunespt) [inline src/sagas/employeeSaga/delete.js]: I see you opted to create the actions object every time you needed to dispatch and action. Do you think this approach could be different?

File: src/shared/components/NavBar/NavBar.jsx [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        <NavLink to='/login' className={classes.link}>Login</NavLink>
+        <LockOpenIcon className={classes.icon}/>
+      </React.Fragment>
+    </React.Fragment>
+  );
+};
+
+const LogoutButton = (props) => {
+  const classes = useStyles();
+  const userName = localStorage.getItem('userLoggedIn') ?
```
Comment (tiagoantunespt) [inline src/shared/components/NavBar/NavBar.jsx:57]: you are accessing to localStorage directly in the component to get the user name. Do you think it's the best approach?

File: src/shared/components/SharedTable/SharedTable.test.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+it('deep snapshot renders correctly', () => {
+  const tree = renderer.create(<SharedTable data={employees}
+                                            headCells={headerCells}
+                                            dialog={dialog}
+                                            dataKey={'id'}
+                                            defaultSortField={'FirstName'}
+                                            defaultSortOrder={'desc'}
+                                            handleSave={onHandleSave}
+                                            handleDeleteClick={onHandleDeleteClick}/>).toJSON();
+  expect(tree).toMatchSnapshot();
```
Comment (tiagoantunespt) [inline src/shared/components/SharedTable/SharedTable.test.js:51]: What do you think about snapshot testing? Do you think it is enough to test this component?

File: src/shared/components/SharedTable/SharedTable.jsx [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+                                       key={headCell.key}>
+                        {getCellBody(headCell, row)}
+                      </StyledTableCell>
+                    ))}
+                  </StyledTableRow>
+                );
+              })}
+          </TableBody>
+        </Table>
+        <TablePagination
```
Comment (tiagoantunespt) [inline src/shared/components/SharedTable/SharedTable.jsx:279]: Imagine this table does not have pagination and the request fetches 1 000 000 employees. What approach would you use here?

**Review by tiagoantunespt — COMMENTED** [review]
Hey Pratibha! I added some questions/comments in you code so we can have a healthy discussion. I think doing code review like this is very good to learn new things :)

Thank you for your time!

File: .idea/codeStyles/codeStyleConfig.xml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+<component name="ProjectCodeStyleConfiguration">
```
Comment (yashikagarg13) [inline .idea/codeStyles/codeStyleConfig.xml:1]: Do you think these files should be committed?

File: src/App.js [hunk 12]
```diff
@@ -1,26 +1,82 @@
...
+    props.logout();
+  };
+
+  useEffect(() => {
+    if (props.token) {
+      setIsLoggedIn(false);
+    }
+
+    if (props.status) {
+      setTimeout(() => {
```
Comment (yashikagarg13) [inline src/App.js:28]: Why we need `setTimeout` here?

File: src/App.js [hunk 13]
```diff
@@ -1,26 +1,82 @@
...
+};
+
+const mapDispatchToProps = (dispatch) => {
+  return {
+    logout: () => dispatch({type: ActionTypes.LOGOUT})
+  };
+};
 
-export default App;
+export default connect(mapStateToProps, mapDispatchToProps)(App);
```
Comment (yashikagarg13) [inline src/App.js:84]: What is the best approach to creat container components?

File: src/actions/employee.action.js [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+export function LoadAllEmployees() {
+  return new Promise(async (resolve, reject) => {
```
Comment (yashikagarg13) [inline src/actions/employee.action.js]: Why would we need `async` here?

File: src/actions/employee.action.js [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    try {
+      if (!employee.id) {
+        employee.id = `5d6f705653106f633ad9${employeesData.length + 1}cde`;
+        employeesData.push(employee);
+      } else {
+        const employeeIndex = employeesData.findIndex(entry => employee.id === entry.id);
+        if (employeeIndex !== -1) {
+          employeesData[employeeIndex] = employee;
+        } else {
+          throw new Error('Employee not found in database');
```
Comment (yashikagarg13) [inline src/actions/employee.action.js]: What could be the best way to define error messages?

File: src/actions/employee.action.js (hunk 14 above)
Comment (yashikagarg13) [reply src/actions/employee.action.js]: Also, I think these calls as dummy API calls. Am I missing on something?

File: src/components/EmployeeDetail/EmployeeDetail.jsx [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+                          onClick={onEditClick}
+                          margin="normal">
+                    Edit
+                  </Button>
+                </div>
+              </div>
+            </div>
+          </Paper>
+        </div>
+      );
```
Comment (yashikagarg13) [inline src/components/EmployeeDetail/EmployeeDetail.jsx:184]: How can we make this file more readable?

File: src/components/EmployeeDetail/EmployeeDetail.jsx [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+    }
+  };
+
+  return (
+    <React.Fragment>
+      {getRedirectHtml()}
```
Comment (yashikagarg13) [inline src/components/EmployeeDetail/EmployeeDetail.jsx:190]: Why would we render the content by this function call and not directly?

File: src/components/EmployeeEdit/EmployeeEdit.jsx [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+  const classes = useStyles();
+  const [cancelClicked, setCancelClicked] = useState(false);
+  const [employee, setEmployee] = useState(initialEmployeeData);
+  const [errorMessage, setErrorMessage] = useState({});
+
+  useEffect(() => {
+    if (props.action === 'edit' && props.employee) {
+      setEmployee(props.employee);
+    }
+  }, [props.action, props.employee, employee]);
```
Comment (yashikagarg13) [inline src/components/EmployeeEdit/EmployeeEdit.jsx:90]: Why would we need `employee` as dependency?

File: src/sagas/index.js [hunk 19]
```diff
@@ -0,0 +1,10 @@
+import { all } from 'redux-saga/effects'
+import {loginFlow} from './loginSaga';
+import {watchFetchEmployees} from './employeeSaga/fetch';
+
+export default function* rootSaga() {
+  yield all([
+    loginFlow(),
+    watchFetchEmployees()
```
Comment (yashikagarg13) [inline src/sagas/index.js]: Where are we using this file?

**Comment by yashikagarg13** [comment]
Hey Pratibha, Superb Job. I added a few questions and comments throughout your code. As Tiago said, let's discuss and learn together.
Also, it would be really nice if you could add some more tests for reducers, sagas, some integration tests for components and containers

File: package.json (hunk 1 above)
Comment (pratibhasahai2301) [reply package.json:12]: I was planning to use it earlier, but then I decided to go with Material UI.
And forgot to remove the dependency.

File: src/components/Login/Login.jsx (hunk 3 above)
Comment (pratibhasahai2301) [reply src/components/Login/Login.jsx:12]: We can use both styles, obviously 1 at a time.
I was just showcasing that we can do it both ways as it was an example project.
The good approach would be to move everything in the css.
Or if one wishes to go with the jss approach, instead of defining style in every file, one can maintain a common file which can have centralized styles.

File: src/reducers/employeeReducer.js (hunk 4 above)
Comment (pratibhasahai2301) [reply src/reducers/employeeReducer.js:47]: We can use 'useReducer' hook in the Component, and define our messages there.

File: src/reducers/employeeReducer.js (hunk 5 above)
Comment (pratibhasahai2301) [reply src/reducers/employeeReducer.js:14]: Because the props are immutable.
And we can not change the initial source.

File: src/reducers/userReducer.js (hunk 6 above)
Comment (pratibhasahai2301) [reply src/reducers/userReducer.js:34]: Currently I am not using any of the stated.
But, generally I use prettier and js linter

File: src/shared/components/NavBar/NavBar.jsx (hunk 8 above)
Comment (pratibhasahai2301) [reply src/shared/components/NavBar/NavBar.jsx:57]: We can create a wrapper around localStorage

File: src/shared/components/SharedTable/SharedTable.test.js (hunk 9 above)
Comment (pratibhasahai2301) [reply src/shared/components/SharedTable/SharedTable.test.js:51]: No, there are lot of functionalities which we can't cater efficiently with Snapshot.
For example, the pagination, sorting and inline editing features which can not be tested with snapshot

File: src/shared/components/SharedTable/SharedTable.jsx (hunk 10 above)
Comment (pratibhasahai2301) [reply src/shared/components/SharedTable/SharedTable.jsx:279]: We can implement virtual table from angular material ui or we can create custom infinite table like feature.

File: .idea/codeStyles/codeStyleConfig.xml (hunk 11 above)
Comment (pratibhasahai2301) [reply .idea/codeStyles/codeStyleConfig.xml:1]: No, i forgot to mention these files in .gitignore

File: src/actions/employee.action.js (hunk 14 above)
Comment (pratibhasahai2301) [reply src/actions/employee.action.js]: Yes, these calls are dummy Api calls. 
I am using async to display asynchronous behavior of the normal Api(s)

File: src/actions/employee.action.js (hunk 15 above)
Comment (pratibhasahai2301) [reply src/actions/employee.action.js]: We can create a constant file to store all the messages.

File: src/components/EmployeeDetail/EmployeeDetail.jsx (hunk 16 above)
Comment (pratibhasahai2301) [reply src/components/EmployeeDetail/EmployeeDetail.jsx:184]: We can use Prettier or js lint to format the files..

File: src/components/EmployeeDetail/EmployeeDetail.jsx (hunk 17 above)
Comment (pratibhasahai2301) [reply src/components/EmployeeDetail/EmployeeDetail.jsx:190]: Because of conditional rendering, I have used this function call.
We can also do inline conditional rendering but that would make the code clumsy and unreadable.

File: src/components/EmployeeEdit/EmployeeEdit.jsx (hunk 18 above)
Comment (pratibhasahai2301) [reply src/components/EmployeeEdit/EmployeeEdit.jsx:90]: console was giving me warning messages, to add the employee dependency

File: src/sagas/index.js (hunk 19 above)
Comment (pratibhasahai2301) [reply src/sagas/index.js]: I was earlier using this file in the store, but i removed its reference. But forgot to remove the file.

File: src/App.js (hunk 13 above)
Comment (pratibhasahai2301) [reply src/App.js:84]: The best approach to define a container component is to have the logic separated from the UI. 
The container should contain all the Logic and pass the data to the UI or View Components to render.

File: src/actions/employee.action.js (hunk 2 above)
Comment (pratibhasahai2301) [reply src/actions/employee.action.js]: Actions are dispatched from the Components. 
Within the respective Action method, we can play around with the data (i.e. we can use any middleware for making an async call to the API(s), or to filter the state data as per the parameters received).
After getting the data, it is the responsibility of action to send the data to the Reducer for manipulating state.

Component -------> Redux Action -------> Reducer -------> set state

File: src/App.js (hunk 12 above)
Comment (pratibhasahai2301) [reply src/App.js:28]: To stabilize the behavior of React Toast

**Comment by pratibhasahai2301** [comment]
Added few tests for reducers, store, and actions
